future/src/ct/ct_splittable.h

```diff
@@ -0,0 +1,203 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <utility>
+
+namespace {
+
+template<typename T>
+std::vector<T> split_items(const T& items) {
+    std::vector<T> items_vect(items.size());
+    for (const auto& item : items) {
+        items_vect.emplace_back(item);
+    }
+    return items_vect;
+}
+
+template<>
+std::vector<Glib::ustring> split_items(const Glib::ustring& str) {
+    std::vector<Glib::ustring> items_vect;
+    for (const auto& ch : str) {
+        items_vect.emplace_back(1, ch);
+    }
+    return items_vect;
+}
+
+}
+template<typename ITEM_T, typename ITEM_T_SPLITTER = decltype(&split_items<ITEM_T>)>
+class CtSplittable
+{
+    using vect_t = std::vector<ITEM_T>;
+public:
+    /**
+     * @brief Swap two CtVectorProxy objects
+     * @param first
+     * @param second
+     */
+    friend void swap(CtSplittable& first, CtSplittable& second) noexcept
+    {
+        using std::swap;
+        
+        swap(first._internal_vec, second._internal_vec);
+        swap(first._item_cache, second._item_cache);
+        swap(first._set_cache, second._set_cache);
+        swap(first._set_vec_cache, second._set_vec_cache);
+    }
+    
+    /**
+     * @brief Construct a new proxy out of iterators compatable with vect_t's ctor
+     * @param begin
+     * @param end
+     */
+    template<class Iterator>
+    CtSplittable(Iterator begin, Iterator end, ITEM_T_SPLITTER item_splitter = &split_items<ITEM_T>) : _item_splitter(item_splitter), _internal_vec(begin, end) {}
+    
+    typename vect_t::reference operator[](typename vect_t::size_type index) {
+        _load_vec_cache(); 
+        return _internal_vec[index]; 
+    }
+    
+    typename vect_t::const_reference operator[](typename vect_t::size_type index) const { 
+        _load_vec_cache();
+        return _internal_vec[index]; 
+    }
+    
+    /**
+     * @brief Implicit covnersion operator
+     * @return
+     */
+    operator ITEM_T() const
+    {
+        _load_cache();
+        return _item_cache;
+    }
+    operator vect_t() const
+    {
+        _load_vec_cache();
+        return _internal_vec;
+    }
+    
+    CtSplittable(const CtSplittable&) = default;
+    CtSplittable(CtSplittable&&) noexcept = default;
+    CtSplittable(ITEM_T item, ITEM_T_SPLITTER item_splitter = &split_items<ITEM_T>) : _item_splitter(item_splitter), _item_cache(std::move(item)), _set_vec_cache(false) {}
+    
+    CtSplittable& operator=(CtSplittable other)
+    {
+        swap(*this, other);
```

Comment:
doesn't look great, = and swap have different meaning, also because of that you don't use const ref

Reply:
This is the copy and swap idiom, [this so post](https://stackoverflow.com/a/3279550/12448530) explains it pretty well

Reply:
Ok, got it